### PR TITLE
ENG-15786: Crash with terminate after 'second parameter' error

### PR DESCRIPTION
**Pull request by Chaozz**
The cause of the problem is that we are trying to create an index based on [STR](https://docs.voltdb.com/UsingVoltDB/sqlfuncstr.php) function.
The second argument of **STR** must be an integer between 0 and 38. An insertion/update of the table will fail if it hit the index and fail to meet the 0~38 constraint.

This fix disallows **STR** in index expressions.

### CODE REVIEW

File: src/frontend/org/voltdb/expressions/AbstractExpression.java [hunk 1]
```diff
@@ -1299,9 +1307,19 @@ public static boolean hasInlineVarType(AbstractExpression expr) {
...
+        final int invalidFunctionId = containsFunctionByIds(INVALID_FUNCTIONS_FOR_IDXS_AND_MVS);
+        if (invalidFunctionId != -1) {
+            if (invalidFunctionId == FunctionSQL.voltGetCurrentTimestampId()) {
+                msg.append("cannot include the function NOW or CURRENT_TIMESTAMP.");
+                return false;
+            } else if (invalidFunctionId == FunctionForVoltDB.FunctionDescriptor.FUNC_VOLT_STR) {
+                msg.append("cannot include the function STR.");
+                return false;
+            } else {
+                msg.append("cannot include invalid function.");
```
Comment (kaikaizi) [inline src/frontend/org/voltdb/expressions/AbstractExpression.java]: `assert false; // cannot reach here`
Also, I'd prefer switch for a bit of clarity.

File: src/frontend/org/voltdb/expressions/AbstractExpression.java [hunk 2]
```diff
@@ -1420,6 +1438,27 @@ private boolean containsFunctionById(int functionId) {
...
     }
 
+    /**
+     * This function will recursively find any function expression with any ID in the functionIds.
+     * If found, return the functionId. Otherwise, return -1.
+     * @param functionIds
+     * @return If found, return the functionId. Otherwise, return -1
+     */
+    private int containsFunctionByIds(Set<Integer> functionIds) {
+        if (this instanceof AbstractValueExpression) {
```
Comment (kaikaizi) [inline src/frontend/org/voltdb/expressions/AbstractExpression.java]: Do you need this here? When?
Or should it belong to `findAllFunctionsSubexpressions()`?

File: tests/frontend/org/voltdb/TestAdhocCreateDropIndex.java [hunk 3]
```diff
@@ -455,7 +455,8 @@ public void testENG15213() throws Exception {
                     Pair.of("CREATE INDEX PI31 ON P5(i) WHERE j > 0;", true),               // partial index with columns and safe predicate on non-empty table: passes
                     Pair.of("CREATE INDEX PI21 ON P5(i, LOG10(j));", false),                 // normal index with unsafe expression on non-empty table: rejected
                     Pair.of("CREATE INDEX PI41 ON P5(i) WHERE LOG(j) > 1 OR j <= 0;", false),// partial index with columns and unsafe predicate on non-empty table: rejected
-                    Pair.of("CREATE INDEX PI51 ON P5(LOG(i)) WHERE LOG(j) > 1 OR j <= 0;", false))  // partial index with unsafe expression and unsafe predicate on non-empty table: rejected
+                    Pair.of("CREATE INDEX PI51 ON P5(LOG(i)) WHERE LOG(j) > 1 OR j <= 0;", false),  // partial index with unsafe expression and unsafe predicate on non-empty table: rejected
+                    Pair.of("CREATE INDEX PI6 ON P5(i) WHERE 'ABC' > STR(i, i);", false))   // STR function is invalid in expressions for indexes
```
Comment (kaikaizi) [inline tests/frontend/org/voltdb/TestAdhocCreateDropIndex.java]: `// .... ENG-15786`

Comment (kaikaizi) [reply tests/frontend/org/voltdb/TestAdhocCreateDropIndex.java]: Also add a test to show that you cannot create index with `STR` function, even if the table is empty. Add tests for `NOW` and `CURRENT_TIMESTAMP` as well.

File: src/frontend/org/voltdb/expressions/AbstractExpression.java (hunk 1 above)
Comment (Chaozz) [reply src/frontend/org/voltdb/expressions/AbstractExpression.java]: Added the assert.

I can not use switch here because `FunctionSQL.voltGetCurrentTimestampId()` is not a constant value(at least the compiler thinks so)

Comment (Chaozz) [reply src/frontend/org/voltdb/expressions/AbstractExpression.java]: Added the assert.

I can not use switch here because `FunctionSQL.voltGetCurrentTimestampId()` is not a constant value(at least the compiler thinks so)

File: src/frontend/org/voltdb/expressions/AbstractExpression.java (hunk 2 above)
Comment (Chaozz) [reply src/frontend/org/voltdb/expressions/AbstractExpression.java]: I use it here https://github.com/VoltDB/voltdb/blob/5b8b8da8b5a0ccc65fd0838bca57ab79ef593ded/src/frontend/org/voltdb/expressions/AbstractExpression.java#L1311

File: tests/frontend/org/voltdb/TestAdhocCreateDropIndex.java (hunk 3 above)
Comment (Chaozz) [reply tests/frontend/org/voltdb/TestAdhocCreateDropIndex.java]: done

File: src/frontend/org/voltdb/expressions/AbstractExpression.java (hunk 1 above)
Comment (kaikaizi) [reply src/frontend/org/voltdb/expressions/AbstractExpression.java]: I see.
You are still missing a false assertion in this else branch?

File: tests/frontend/org/voltdb/TestAdhocCreateDropIndex.java [hunk 4]
```diff
@@ -450,12 +450,20 @@ public void testENG15213() throws Exception {
                     Pair.of("CREATE INDEX PI4 ON P5(i) WHERE LOG(j) > 1 OR j <= 0;", true), // partial index with columns and unsafe predicate on empty table: passes
                     Pair.of("CREATE INDEX PI5 ON P5(LOG(i)) WHERE LOG(j) > 1 OR j <= 0;", true), // partial index with unsafe expression and unsafe predicate on empty table: passes
                     Pair.of("DROP INDEX PI2; DROP INDEX PI4; DROP INDEX PI5", true),        // clean up indexes with unsafe operations
-                    Pair.of("INSERT INTO P5 values(0, 0);", true),                          // Table has rows: passes
```
Comment (kaikaizi) [inline tests/frontend/org/voltdb/TestAdhocCreateDropIndex.java]: Please keep this line (and move below L457). I needed this to test `LOG(j)` on index.

File: src/frontend/org/voltdb/expressions/AbstractExpression.java [hunk 5]
```diff
@@ -1316,7 +1334,7 @@ public boolean isValidExprForIndexesAndMVs(StringBuffer msg, boolean isMV) {
             }
             return false;
         } else if (hasUserDefinedFunctionExpression()) {
-            msg.append("cannot contain calls to user defined functions.");
```
Comment (kaikaizi) [inline src/frontend/org/voltdb/expressions/AbstractExpression.java]: I think we need to keep this message?

File: src/frontend/org/voltdb/expressions/AbstractExpression.java [hunk 6]
```diff
@@ -1316,7 +1334,7 @@ public boolean isValidExprForIndexesAndMVs(StringBuffer msg, boolean isMV) {
...
+            assert false; // cannot reach here
```
Comment (kaikaizi) [inline src/frontend/org/voltdb/expressions/AbstractExpression.java]: I think we might reach here: have a UDF in index.
